### ISSUE TITLE
apache module can't do ip-based VirtualHosts

### ISSUE THREAD

**Issue by JasonWoof** [NixOS/nixpkgs#187645]
### Describe the bug
This commit https://github.com/NixOS/nixpkgs/commit/79215f0df1ddf4bf0db7dc4c5789f8dae9f9bb02#diff-c4e76be6d8d687a02103da358697dd78a98738f3a1bc20c52b3a00fbccf2e6a9R262

Changes apache-httpd/default.nix so that it generates ``Listen`` directives from all the virtualHosts.

This makes it impossible to have two services.httpd.virtualHosts with different IP addresses and the same port, eg:

<VirtualHost 127.0.0.1:80>...``

and

``<VirtualHost *:80>...``

With virtual hosts like this you'd need this Listen directive:

``Listen *:80``

But the apache nix module generates:

```
Listen 127.0.0.1:80
Listen *:80
```

This is an invalid configuration, and Apache will refuse to start, because it's not possible to listen on both of these at once. You get errors like:

``(98)Address already in use: AH00072: make_sock: could not bind to address 0.0.0.0:80``

### Steps To Reproduce

Something like this:

```
services.httpd.virtualHosts = {
    "1localhost" = {
        listen = [
            {ip = "127.0.0.1"; port = 80; ssl = false;}
            {ip = "[::1]"; port = 80; ssl = false;}
        ];
        addSSL = false;
        extraConfig = "VirtualDocumentRoot /var/www/private/%-2+"
    }:
    "2public" = {
        # no listen directive
        addSsl = false;
        extraConfig = "VirtualDocumentRoot /var/www/public/%0"
    };
};
```

### Expected behavior
It should be possible to do ip/port virutalhosts

Possible solutions:
1. The code to generate ``Listen`` directives is improved
2. There is a way to specify the ``Listen`` directives for cases like this where the auto-generated ones don't work.

### Notify maintainers
I don't understand how to look up meta.maintainers. Please add a shell one-liner to the instructions here like you do for Metadata below.

I think @aanderse commited the change that broke this use case.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.19.1, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(jasonwoof): `"nixos-21.11pre317242.bc9b956714e, nixpkgs-21.11pre317548.e0ce3c683ae"`
 - channels(root): `"nixos-22.05, nixos-unstable"`
 - nixpkgs: `/home/jasonwoof/.nix-defexpr/channels/nixpkgs`
```

**Comment by aanderse** [comment]
Specify a `listen` on each virtual host and this will allow you to bind to multiple ip addresses, across multiple virtual hosts if desired, to the same port. I used to do this all the time.

**Comment by JasonWoof** [comment]
I think I figured out why this works in some configurations, but not mine. Some I think it has to do with how the ip addresses land on devices, or maybe just catchall addresses conflict with more specific ones.

If I have one VirtualHost listening on 127.0.0.1 and one listening on a static IP on a different network device, then it works fine, but one on ``127.0.0.1`` and one on ``*`` fails (see below.)

The host I'm configuring gets one of it's IP addresses from DHCP, so it'd be useful to be able to have a VirtualHost on ``*``.

Here's a minimal version of what I'm trying to do that exhibits the bug:

```
    services.httpd.virtualHosts = {
        "1_localhost" = {
            hostName = "localhost";
            documentRoot = "/var/www/private";
            listen = [{ip = "127.0.0.1"; port = 80; ssl = false;}];
            addSSL = false;
        };
        "2_public" = {
            hostName = "public";
            documentRoot = "/var/www/public";
            # adding the following line will avoid this bug:
            # listen = [{ip = "200:1234:1234:1234:1234:1234:1234:1234"; port = 80; ssl = false;}];
            addSSL = false;
        };
    };
```

this creates the following in /etc/httpd/httpd.conf

```
Listen 127.0.0.1:80 http
Listen *:80 http
```

and apache fails to start with this error message:

```
Aug 21 12:19:10 gold httpd[9950]: (98)Address already in use: AH00072: make_sock: could not bind to address [::]:80
Aug 21 12:19:10 gold httpd[9950]: (98)Address already in use: AH00072: make_sock: could not bind to address 0.0.0.0:80
```

I don't know what makes certain IPs conflict, but it seems like ``*`` and ``0.0.0.0`` both conflict with ``127.0.0.1``.

If it has nothing to do with which card the ip addresses are hosted on, and just that the catchall addresses conflict with specific ones, then this could be fixed by making the dedup code that generates the ``Listen`` directives smarter. i.e. making it delete any ipv4 addresses if there's ``0.0.0.0`` or ``*``, and delete any ipv6 addresses if there's a ``[::]`` or ``*``.

Or, if it's not as simple as that, then the apache nix module could go back to allowing configuration.nix to supply the ``Listen`` directives directly.

**Comment by aanderse** [comment]
I'm having trouble understanding this. As far as I can recall `apache` doesn't allow this - it has nothing to do with this NixOS module. Once you mention a specific ip address:port for one virtual host you have to mention a specific ip address for all virtual hosts.

Can you please post the *valid* `apache` configuration you would write if you weren't using NixOS? Stick to ipv4 only to keep things simple please.

**Comment by JasonWoof** [comment]
apache _does_ allow virtualhost's ip/port pairs to overlap in whatever ways. Here's (simplified) what I typically do on my laptop:

```
Listen *:80
<VirtualHost 127.0.0.1:80>
    VirtualDocumentRoot [map *.localhost to ~/dev/*/]
</VirtualHost>
<VirtualHost *:80>
    VirtualDocumentRoot [map *.whatever to ~/dev/public/*
</VirtualHost>
```
This works beautifully. I have access each ``~/dev/*/`` by going to ``http://DIRNAME.localhost``, but other machines on the network can only access ``~/dev/public/*/``

The ``Listen`` directives clash in the way that you understand (apache won't run), but the listen-like values that appear  ``<VirtualHost [HERE]>`` can overlap in a myriad of ways and apache is fine with it, and chooses the correct VirtualHost in my experience.

**Comment by aanderse** [comment]
Right. Sorry, I recall now. I haven't touched `apache` in a while now but this is coming back to me.

I'll give this some thought.

**Comment by aanderse** [comment]
After some thought I don't see a better solution than your proposal to add an optional `services.httpd.listen` option which can override the default.

Maybe someone can come to with some reasonable logic to pick better defaults... but I don't use `apache` that much anymore, so I'm out.

**Comment by JasonWoof** [comment]
OK, Thanks for circling back.

Maybe I'll write something up in the coming months. I think I know what the logic should be.
